### PR TITLE
Implement week picker component

### LINKED ISSUE #2033: Can we have a week picker ?
I see month picker , year picker but no week picker :) .

### PR DESCRIPTION
**Pull request by TiesWestendorp**
fixes #2033

Introduces the `showWeekPicker` setting, which allows picking a week. Internally, the date representing the week is the first day of that week. It's important that the `dateFormat` prop matches the `locale` in the sense that they have the same starting day of the week. In the same vein, the `selected` prop should refer to a date that's at the beginning of a week.

Peek 2021-03-01 16-15

**Question 1**: should left and right arrows navigate the months, rather than weeks? **Answer:** no.
**Question 2**: should the week component be considered selected/keyboard-selected, and the styling be derived from that, or should the days be considered selected/keyboard-selected like they currently are?

### CODE REVIEW

**Comment by TiesWestendorp** [comment]
@martijnrusschen or someone else, can you take a look at this PR, and give your take on the questions I posed above?

**Comment by tlehtimaki** [comment]
Q2: I think the selecting of all dates in the week seems logical, but then I think you should not be able to individually select the dates.
Q1: For how its implemented now the up/down arrows are logical for browsing the week. The browsing of month suddenly with left/right arrow might be confusing for the user, if user is used to browsing days with those keys.

**Comment by TiesWestendorp** [comment]
> Q2: I think the selecting of all dates in the week seems logical, but then I think you should not be able to individually select the dates.

My question was actually more geared towards the internal state; each day component has a 'selected' and 'keyboard-selected' state, which I reuse for the week picker here. Another way would be to not consider these days selected or keyboard-selected in this case, but the week component that wraps these days. There's a bit of a trade-off here: the former option borrows heavily from the existing styling, meaning it doesn't introduce much complexity in the stylesheets, but is potentially less flexible. As to your comment: clicking on a particular date will currently select the corresponding date - do you mean to say that you don't agree with this? My issue with that is that when the week numbers are not present you won't have a meaningful thing to click on. A takeaway from this may be that the week numbers should always be shown when the week picker is used, but that is also currently not the case.

> Q1: For how its implemented now the up/down arrows are logical for browsing the week. The browsing of month suddenly with left/right arrow might be confusing for the user, if user is used to browsing days with those keys.

I agree, but wanted to lay the options on the table. :+1: Currently left/right behave the same as up/down in the week picker.

**Comment by tlehtimaki** [comment]
I understood that the concept would be to select just weeks, so I thought there would be no need to select a single date, but here that single date would convert in to the week that the date belongs to.

**Comment by TiesWestendorp** [comment]
> I understood that the concept would be to select just weeks, so I thought there would be no need to select a single date, but here that single date would convert in to the week that the date belongs to.

Correct, just having the week number select the first date of a week was already possible by using a callback on the week number. What this PR introduces is the additional logic required to properly display the state, changed keyboard navigation to match and a fallback that ensures the first day of a week is always used.

**Comment by TiesWestendorp** [comment]
Could some one take a look at why the specs are failing? I'm not really sure what's going wrong.

**Comment by martijnrusschen** [comment]
Can you try to pull the latest master in? I think the issue was fixed
earlier.

**Comment by TiesWestendorp** [comment]
The failing spec seems to be fixed by rebasing, but I'm still getting errors during linting...

**Comment by martijnrusschen** [comment]
It looks like something happened with the rebase, seeing some changes in the PR that already got merged. Regarding the lint errors, it seems they're related to the changes in this PR.

**Comment by martijnrusschen** [comment]
Checking in, can you merge with the latest master so we can release this functionality?

**Comment by TiesWestendorp** [comment]
> Checking in, can you merge with the latest master so we can release this functionality?

Did so just now, that at least cleans up the git history. :+1: I'm still not sure about the lint errors though. :shrug:

**Comment by mykolap** [comment]
And proper format string for year/week will also be supported for this peeker? Currently we are forced to use literals "ww/YYY", with 3 letters Y, instead of YY or YYYY, because date-fns rejects ww/YYYY or ww/YY. And week should be used only with big Y, not small y. See https://github.com/Hacker0x01/react-datepicker/issues/2661. Week peeker without support of WeekYear tokens in date-fns will be not complete.

**Comment by TiesWestendorp** [comment]
> And proper format string for year/week will also be supported for this peeker? Currently we are forced to use literals "ww/YYY", with 3 letters Y, instead of YY or YYYY, because date-fns rejects ww/YYYY or ww/YY. And week should be used only with big Y, not small y. See #2661. Week peeker without support of WeekYear tokens in date-fns will be not complete.

I'm not really familiar with the problem you're describing. The only thing that's important for this PR, is that `date-fns` is able to parse the the given format, and that in doing so it returns a date that corresponds to the first calendar cell in a given week. In the given example this means I've used the format `I/R`.

**Comment by mykolap** [comment]
To show start of week correctly - you should pass locale to dateFns, after that if locale German - you will get Monday as a start, if US - Sunday, but issue is with year, because year for week and calendar year - are different, and only hack with 3 letters Y is working when we want to handle year correcly... And it's very ugly hack...

**Comment by TiesWestendorp** [comment]
> To show start of week correctly - you should pass locale to dateFns, after that if locale German - you will get Monday as a start, if US - Sunday, but issue is with year, because year for week and calendar year - are different, and only hack with 3 letters Y is working when we want to handle year correcly... And it's very ugly hack...

I would say that this seems like a separate issue from what's being done in this PR. _Given_ that `format` and `parse` work correctly, the changes from this PR work correctly.

**Comment by TiesWestendorp** [comment]
@martijnrusschen I've fixed the lint errors. Could you take a look at this PR?

File: src/day.jsx [hunk 1]
```diff
@@ -259,10 +270,17 @@ export default class Day extends React.Component {
   getTabIndex = (selected, preSelection) => {
     const selectedDay = selected || this.props.selected;
     const preSelectionDay = preSelection || this.props.preSelection;
-
+    // TODO: use isStartOfWeek
```
Comment (martijnrusschen) [inline src/day.jsx:273]: Is this still something you want to change here?

Comment (TiesWestendorp) [reply src/day.jsx:273]: I think I just forgot to remove a comment there.

**Comment by TiesWestendorp** [comment]
@martijnrusschen Could you still have a look at this PR?

**Comment by martijnrusschen** [comment]
This looks good! Can you resolve the merge conflict so I can merge it in? I'm preping a major release for this week and would love to get this in.

**Comment by TiesWestendorp** [comment]
:+1: Done.

**Comment by martijnrusschen** [comment]
I ran the tests, and it looks like there's some things breaking. Can you take a look?

**Comment by TiesWestendorp** [comment]
Should be fixed now.

File: test/week_picker_test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import React from "react";
```
Comment (martijnrusschen) [inline test/week_picker_test.js:1]: Thinking a bit on code coverage.. Might be good to add some tests that checks the behavior of showing/hiding the weekpicker as well. Making sure the weekpicker isn't shown when the prop isn't passed in and shown when it is passed in.

Comment (martijnrusschen) [reply test/week_picker_test.js:1]: Codecov added some comments on coverage as well, would be great to improve the coverage to make sure we don't break this in the future.

**Review by martijnrusschen — CHANGES_REQUESTED** [review]
Let's improve the test coverage to make sure we won't break this in the future.

**Comment by georgecosmin98** [comment]
Hello!

Any news about week-picker? When can we expect a release?

**Comment by johnnyxbell** [comment]
Any news on when we can get this merged?

**Comment by sffirm** [comment]
Why is it taking so long?

**Comment by manzoorwanijk** [comment]
Any update on this?

**Comment by nacimgoura** [comment]
When will this feature be merged?

**Comment by arikm66** [comment]
I really need this feature :)  When will it be merged?

**Comment by martijnrusschen** [comment]
Perhaps someone can take a look and bring this to the finish line.

**Comment by Seaviello** [comment]
Hi,
we are also currently interested in that feature. Is there something that outsider can do to help you out with making this happen or right now it's mostly in the reviewers hand?

**Comment by akashraj9828** [comment]
Any update on this... really need this feature. Thanks in advance.

**Comment by akashraj9828** [comment]
Not sure if anyone needs this anymore but I was able to implement the week picker with the existing library... 
Demo: 
https://codesandbox.io/s/react-datepicker-example-forked-89fn7?file=/src/WeekPicker.tsx

**Comment by ngajhede** [comment]
Hey @correahi! Thanks for this. I was wondering how you display the `startDate to endDate` in the input box?

**Comment by correahi** [comment]
> Hey @correahi! Thanks for this. I was wondering how you display the `startDate to endDate` in the input box?

I don't have access to that codebase anymore but i'm pretty sure it's just a span with some text inside, something like `${startOfISOWeek(selectedDate)} to ${endOfISOWeek(selectedDate)}` and `onClick` it opens the datepicker, it's not the default input.

**Comment by gtaylor44** [comment]
Will this be merged into master?

**Pull request closed without merging**